File: credit.c
#include <cs50.h>
#include <stdio.h>

int countDigits(long number);
bool isVisa(long number);
bool isMasterCard(long number);
bool isAmericanExpress(long number);
bool calculateChecksum(long number);

int main(void)
{
    // Get the card number to validate
    long number = get_long("Number: ");
    // Count the credit card digits number
    int digits = countDigits(number);

    // If checksum is valid verify cards
    if (calculateChecksum(number))
    {
        // Checking for VISA Card
        if (isVisa(number) && (digits == 13 || digits == 16))
        {
            printf("VISA\n");
        }
        // Checking for American Express Card
        else if (isAmericanExpress(number) && digits == 15)
        {
            printf("AMEX\n");
        } 
        // Checking for MasterCard Card
        else if (isMasterCard(number) && digits == 16)
        {
            printf("MASTERCARD\n");
        }
        // Card number not matching any card pattern
        else
        {
            printf("INVALID\n");
        }
    }
    // Checksum don't pass
    else
    {
        printf("INVALID\n");
    }
}

// Method to count the digits in the number
int countDigits(long number)
{
    int count = 0;
    while (number != 0)
    {
        number = number / 10;
        count++;
    }
    return count;
}

// For VISA cards
bool isVisa(long number)
{
    int digits = countDigits(number);
    int start = 0;
    if (digits == 13)
    {
        start = number / 1000000000000;
    }
    else if (digits == 16)
    {
        start = number / 1000000000000000;
    }
    
    return start == 4;
}

// For MasterCard
bool isMasterCard(long number)
{
    int digits = countDigits(number);
    int start = number / 100000000000000;
    return start == 51 || start == 52 || start == 53 || start == 54 || start == 55;
}

// For AmericanExpress
bool isAmericanExpress(long number)
{
    int digits = countDigits(number);
    int start = number / 10000000000000;
    return start == 34 || start == 37;
}

// Method to calculate the checsum
bool calculateChecksum(long number)
{
    int sum = 0;
    long number1 = number;
    int digits = countDigits(number);

    bool analize = false;
    for (int i = 0; i < digits; i++)
    {
        // Get alternated numbers
        if (analize == true)
        {
            long num = number1 % 10;
            int multiplication = num * 2;
            // If the multiplication is greather than 10, sum the digits
            if (multiplication >= 10)
            {
                sum += multiplication % 10;
                multiplication = multiplication / 10;
                sum += multiplication;
            }
            else
            {
                sum += multiplication;
            }
        }
        number1 = number1 / 10;
        analize = !analize;
    }

    analize = true;
    number1 = number;
    for (int i = 0; i < digits; i++)
    {
        if (analize)
        {
            int num2 = number1 % 10;
            sum += num2;
        }

        number1 = number1 / 10;
        analize = !analize;
    }

    return sum % 10 == 0;
}

File: credit.py
from cs50 import get_int

# main function
def main():
    # get the card number to validate
    number = get_int("Number: ")

    # count the credit card digits number
    digits = countDigits(number)

    # if checksum is valid verify cards
    if calculateChecksum(number):
        # checking for VISA Card
        if isVisa(number) and (digits == 13 or digits == 16):
            print("VISA")
        # Checking for American Express Card
        elif isAmericanExpress(number) and digits == 15:
            print("AMEX")
        # Checking for MasterCard Card
        elif isMasterCard(number) and digits == 16:
            print("MASTERCARD")
        # Card number not matching any card pattern
        else:
            print("INVALID")
    # Checksum don't pass
    else:
        print("INVALID")


# method to count number of digits in the credit card number
def countDigits(number):
    count = 0

    while number != 0:
        number = number // 10
        count += 1

    return count

# method to calculate the checsum
def calculateChecksum(number):
    # initialize the variables to be used
    sum = 0
    number1 = number
    digits = countDigits(number)
    analize = False

    for i in range(digits):
        # get alternated numbers
        if analize:
            num = number1 % 10
            multiplication = num * 2

            # if the multiplication is greather than 10, sum the digits
            if multiplication >= 10:
                sum += multiplication % 10
                multiplication = multiplication // 10
                sum += multiplication
            else:
                sum += multiplication;

        number1 = number1 // 10
        analize = not analize

    analize = True
    number1 = number

    for i in range(digits):
        if analize:
            num2 = number1 % 10
            sum += num2

        number1 = number1 // 10
        analize = not analize

    return sum % 10 == 0

# For VISA cards
def isVisa(number):
    digits = countDigits(number)
    start = 0
    if digits == 13:
        start = number // 1000000000000
    elif digits == 16:
        start = number // 1000000000000000

    return start == 4

# For AmericanExpress
def isAmericanExpress(number):
    digits = countDigits(number)
    start = number // 10000000000000
    values = [34, 37]
    return start in values

# For MasterCard
def isMasterCard(number):
    digits = countDigits(number)
    start = number // 100000000000000
    values = [51, 52, 53, 54, 55]
    return start in values


main()
